include/SH3/system/input.hpp

```diff
@@ -0,0 +1,533 @@
+/** @file
+ *  Defines the @ref sh3::system::input_system.
+ */
+#ifndef SH3_SYSTEM_INPUT_HPP_INCLUDED
+#define SH3_SYSTEM_INPUT_HPP_INCLUDED
+
+#include <bitset>
+#include <chrono>
+#include <cstdint>
+#include <ratio>
+#include <string>
+
+#include <boost/container/flat_map.hpp>
+#include <boost/operators.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+
+#include <SDL2/SDL_keyboard.h>
+#include <SDL2/SDL_mouse.h>
+#include <SDL2/SDL_timer.h>
+
+#include "input_config.hpp"
+
+struct vector2
+{
+    double x, y;
+};
+
+namespace sh3 { namespace system {
```

Comment:
I'll have to go through and change the namespaces so they conform to this. At this point I'm not sure how much difference there is between SH2 and SH3, but it's probably best we prepare early.